Client/game_sa/CDynamicPool.h

```diff
@@ -0,0 +1,171 @@
+};
+
+template <std::size_t InitialSize>
+struct PoolGrownByHalfStrategy
+{
+    static constexpr std::size_t GetInitialSize() { return InitialSize; }
+    static constexpr std::size_t GetNextSize(std::size_t index) { return InitialSize / 2; }
+};
+
+template <typename PoolObjT, typename GrownStrategy>
+class CDynamicPool
+{
+public:
+    CDynamicPool()
+    {
+        constexpr size_t initialSize = GrownStrategy::GetInitialSize();
+        m_poolParts.emplace_back(std::make_unique<CDynamicPoolPart<PoolObjT>>(initialSize));
+    }
+
+    PoolObjT* AllocateItem()
+    {
+        for (auto& pool : m_poolParts)
+        {
+            if (pool->HasFreeSize())
+                return pool->AllocateItem();
+        }
+
+        return AllocateNewPart()->AllocateItem();
+    }
+
+    void RemoveItem(PoolObjT* item)
+    {
+        for (auto& pool : m_poolParts)
+        {
+            if (pool->OwnsItem(item))
+            {
+                pool->RemoveItem(item);
+                return;
+            }
+        }
+
+        throw std::exception("Invalid item for CDynamicPool::RemoveItem");
+    }
+
+    std::size_t GetCapacity() const noexcept
+    {
+        std::size_t size = 0;
+        for (auto& pool : m_poolParts)
+            size += pool->GetCapacity();
+
+        return size;
+    }
+
+    std::size_t GetUsedSize() const noexcept
+    {
+        std::size_t size = 0;
+        for (auto& pool : m_poolParts)
+            size += pool->GetUsedSize();
+
+        return size;
+    }
+
+    bool SetCapacity(std::size_t newSize) {
+        if (newSize == 0)
+            return false;
+
+        std::size_t currentSize = GetCapacity();
+
+        if (currentSize == newSize)
+            return false;
+        else if (currentSize < newSize)
+        {
+            // Grown
+            while (currentSize < newSize)
+            {
+                try
+                {
+                    auto* nextPart = AllocateNewPart();
+                    currentSize += nextPart->GetCapacity();
+                }
+                catch (const std::bad_alloc&)
+                {
+                    return false;
+                }
+            }
+        }
+        else
+        {
+            // Shrink
+            while (true)
+            {
+                auto* part = m_poolParts.back().get();
+                if (part->GetUsedSize() != 0)
+                    return false;
+
+                currentSize -= part->GetCapacity();
+                if (currentSize < newSize)
+                    return false;
+
+                m_poolParts.pop_back();
+
+                if (currentSize == newSize)
+                    return true;
+            }
+        }
+        
+        return true;
+    }
+
+private:
+    CDynamicPoolPart<PoolObjT>* AllocateNewPart()
+    {
+        const std::size_t nextSize = GrownStrategy::GetNextSize(m_poolParts.size());
+        m_poolParts.emplace_back(std::make_unique<CDynamicPoolPart<PoolObjT>>(nextSize));
```

Comment:
iirc `emplace_back` returns an iterator to the item

Reply:
It returns void in C++ 14